src/charm.py

```diff
@@ -259,14 +317,21 @@ def _on_admin_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app's admin ingress URL: %s", event.url)
 
+        self._send_relation_info()
+
     def _on_public_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app's public ingress URL: %s", event.url)
 
+        self._update_config_restart_service(event)
+
     def _on_ingress_revoked(self, event: IngressPerAppRevokedEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app no longer has ingress")
 
+        self._update_config_restart_service(event)
+        self._send_relation_info()
```

Comment:
If something goes wrong in `_update_config_restart_service` (e.g. can't connect to the container or the db is not ready), the `_send_relation_info` method will still run and add the URLs in the databag.

This is not necessarily the wrong behavior, but we should be careful about it. 

Placing the URLs in the databag is meant to signal to Kratos that Hydra is ready or just that WHEN Hydra is up and ready that is where it will be found?

Reply:
I meant the second option you described: tell kratos that hydra will be available under this URL when ready. I also do it when there's no need to run `_update_config_restart_service` (for example `_on_hydra_relation_changed`). Hydra will then push its new endpoints whenever they get updated.
This is to avoid complexity - I think it's unnecessary to block kratos when the `hydra:kratos` relation is in place but hydra is not ready.
I also changed the default url to a predictable endpoint.